// ai-token-counter/index.js
"use strict";

const CJK_PATTERN =
  /[\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uf900-\ufaff]/g;
const WORD_PATTERN = /[A-Za-z]+(?:'[A-Za-z]+)*/g;
const NUMBER_PATTERN = /\d+(?:[.,:/-]\d+)*/g;
const PUNCTUATION_PATTERN = /[!-/:-@[-`{-~]/g;
const EMOJI_PATTERN =
  /[\u{1f300}-\u{1f5ff}\u{1f600}-\u{1f64f}\u{1f680}-\u{1f6ff}\u{1f900}-\u{1f9ff}\u{1fa70}-\u{1faff}]/gu;

function detectProvider(model) {
  if (typeof model !== "string" || model.trim() === "") {
    throw new TypeError("model must be a non-empty string");
  }

  const normalizedModel = model.toLowerCase();
  const compactModel = normalizedModel.replace(/[\s_/.:]+/g, "-");

  if (
    compactModel.includes("gpt") ||
    compactModel.includes("o1") ||
    compactModel.includes("o3") ||
    compactModel.includes("o4") ||
    compactModel.includes("text-embedding") ||
    compactModel.includes("openai")
  ) {
    return "openai";
  }

  if (
    compactModel.includes("claude") ||
    compactModel.includes("anthropic") ||
    compactModel.includes("haiku") ||
    compactModel.includes("sonnet") ||
    compactModel.includes("opus")
  ) {
    return "claude";
  }

  throw new RangeError(
    "Unsupported model. Use an OpenAI or Claude model name."
  );
}

function countMatches(text, pattern) {
  const matches = text.match(pattern);
  return matches ? matches.length : 0;
}

function countWordTokens(text) {
  const words = text.match(WORD_PATTERN) || [];
  let total = 0;

  for (const word of words) {
    const length = word.length;

    if (length <= 4) {
      total += 1;
      continue;
    }

    total += Math.ceil(length / 4);

    if (/[A-Z]/.test(word) && /[a-z]/.test(word)) {
      total += 0.15;
    }
  }

  return total;
}

function estimateBaseTokens(text) {
  const cjkCount = countMatches(text, CJK_PATTERN);
  const punctuationCount = countMatches(text, PUNCTUATION_PATTERN);
  const numberGroups = countMatches(text, NUMBER_PATTERN);
  const emojiCount = countMatches(text, EMOJI_PATTERN);
  const whitespaceGroups = countMatches(text, /\s+/g);
  const wordTokens = countWordTokens(text);
  const visibleChars = text.replace(/\s/g, "").length;
  const latinLikeChars = Math.max(
    0,
    visibleChars - cjkCount - punctuationCount - emojiCount
  );
  const residualChars = Math.max(0, latinLikeChars - countMatches(text, WORD_PATTERN));

  return {
    cjkCount,
    punctuationCount,
    numberGroups,
    emojiCount,
    whitespaceGroups,
    wordTokens,
    residualChars
  };
}

function countTokens(text, model) {
  if (typeof text !== "string") {
    throw new TypeError("text must be a string");
  }

  if (text.length === 0) {
    return 0;
  }

  const provider = detectProvider(model);
  const {
    cjkCount,
    punctuationCount,
    numberGroups,
    emojiCount,
    whitespaceGroups,
    wordTokens,
    residualChars
  } = estimateBaseTokens(text);

  const openAiEstimate =
    wordTokens +
    cjkCount * 1.55 +
    numberGroups * 0.9 +
    punctuationCount * 0.33 +
    whitespaceGroups * 0.12 +
    emojiCount * 2.2 +
    residualChars * 0.25 +
    0.5;

  const claudeEstimate =
    wordTokens * 0.96 +
    cjkCount * 1.4 +
    numberGroups * 0.85 +
    punctuationCount * 0.28 +
    whitespaceGroups * 0.1 +
    emojiCount * 2 +
    residualChars * 0.22 +
    0.35;

  const estimate = provider === "openai" ? openAiEstimate : claudeEstimate;

  return Math.max(1, Math.ceil(estimate));
}

function countMessages(messages, model) {
  if (!Array.isArray(messages)) {
    throw new TypeError("messages must be an array");
  }

  if (messages.length === 0) {
    return 0;
  }

  detectProvider(model);

  let total = 0;

  for (const message of messages) {
    if (!message || typeof message !== "object") {
      throw new TypeError("each message must be an object");
    }

    const role = typeof message.role === "string" ? message.role : "";
    const content = message.content;

    if (typeof content !== "string") {
      throw new TypeError("each message content must be a string");
    }

    // Chat payloads include some structure beyond raw text, so add a small
    // overhead for each message plus a tiny role cost.
    total += countTokens(content, model);
    total += 3;

    if (role) {
      total += Math.max(1, Math.ceil(role.length / 8));
    }

    if (message.name && typeof message.name === "string") {
      total += Math.max(1, Math.ceil(message.name.length / 8));
    }
  }

  return total;
}

module.exports = {
  countTokens,
  countMessages
};
